File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        HashMap<Node , Node> mp = new HashMap<>();
        mp.put(node , new Node(node.val));
        Queue<Node> q = new LinkedList<>();
        q.offer(node);
        while(!q.isEmpty()){
            Node root = q.peek();q.poll();
            for(Node e : root.neighbors){
                if(!mp.containsKey(e)){
                    mp.put(e , new Node(e.val));
                    q.offer(e);
                }
                mp.get(root).neighbors.add(mp.get(e));
            }
        }
        return mp.get(node);
    }
}

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(!node) return NULL;
        unordered_map<Node* , Node*> mp;
        queue<Node*> q;
        q.push(node);
        mp[node] = new Node(node->val);
        while(!q.empty()){
            Node* root = q.front();q.pop();
            for(auto& e : root->neighbors){
                if(mp.find(e) != mp.end()){
                    mp[root]->neighbors.push_back(mp[e]);
                }else{
                    mp[e] = new Node(e->val);
                    mp[root]->neighbors.push_back(mp[e]);
                    q.push(e);
                }
            }
        }
        return mp[node];
    }
};